crates/task-center/src/lib.rs

```diff
@@ -0,0 +1,554 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use futures::FutureExt;
+use restate_types::identifiers::PartitionId;
+use restate_types::tasks::{TaskId, TaskKind};
+use std::collections::HashMap;
+use std::panic::AssertUnwindSafe;
+use std::sync::atomic::{AtomicBool, AtomicI32, AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+use futures::Future;
+use tokio::task::JoinHandle;
+use tokio::task_local;
+use tokio_util::sync::{CancellationToken, WaitForCancellationFutureOwned};
+use tracing::{debug, error, info, instrument, warn};
+
+static NEXT_TASK_ID: AtomicU64 = AtomicU64::new(0);
+const EXIT_CODE_FAILURE: i32 = 1;
+
+#[derive(Debug, thiserror::Error)]
+#[error("system is shutting down")]
+pub struct ShutdownError;
+
+/// Used to create a new task center. In practice, there should be a single task center for the
+/// entire process but we might need to create more than one in integration test scenarios.
+pub struct TaskCenterFactory {}
+impl TaskCenterFactory {
+    pub fn create(runtime: tokio::runtime::Handle) -> TaskCenter {
+        TaskCenter {
+            inner: Arc::new(TaskCenterInner {
+                runtime,
+                global_cancel_token: CancellationToken::new(),
+                shutdown_requested: AtomicBool::new(false),
+                current_exit_code: AtomicI32::new(0),
+                tasks: Mutex::new(HashMap::new()),
+            }),
+        }
+    }
+}
+
+/// Task center is used to manage long-running and background tasks and their lifecycle.
+#[derive(Clone)]
+pub struct TaskCenter {
+    inner: Arc<TaskCenterInner>,
+}
+
+static_assertions::assert_impl_all!(TaskCenter: Send, Sync, Clone);
+
+impl TaskCenter {
+    /// Used to monitor an on-going shutdown when requested
+    pub fn watch_shutdown(&self) -> WaitForCancellationFutureOwned {
+        self.inner.global_cancel_token.clone().cancelled_owned()
+    }
+
+    /// The exit code that the process should exit with.
+    pub fn exit_code(&self) -> i32 {
+        self.inner.current_exit_code.load(Ordering::Acquire)
+    }
+
+    /// Triggers a shutdown of the system. All running tasks will be asked gracefully
+    /// to cancel within a given timeout.
+    #[instrument(level = "error", skip(self, exit_code))]
+    pub async fn shutdown_node(&self, reason: &str, exit_code: i32) {
+        let inner = self.inner.clone();
+        if inner
+            .shutdown_requested
+            .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
+            .unwrap_or_else(|e| e)
+        {
+            // already shutting down....
+            return;
+        }
+        let start = Instant::now();
+        inner.current_exit_code.store(exit_code, Ordering::Release);
+
+        if exit_code != 0 {
+            info!("** Shutdown requested");
+        } else {
+            warn!("** Shutdown requested");
+        }
```

Comment:
Should we warn if the exit code is non-zero?

```suggestion
        if exit_code == 0 {
            info!("** Shutdown requested");
        } else {
            warn!("** Shutdown requested");
        }
```